### PR TITLE
[DO NOT MERGE] Clan Support

### PR DESCRIPTION
**Pull request by infiniterasadev**
**DO NOT MERGE**

This is a WIP branch for clan support based on `dev-creatures`. I'm pull requesting so that I can get some eyes on this before I jump too far in. 

Please review and share your feedback on correctness. I will continue to push to this PR as I implement more and we can merge when it's complete (feel free to add a do not merge label)

### CODE REVIEW

**Comment by infiniterasadev** [comment]
@Blumster @krssrb Please take a peek at this when you can. I want to make sure I'm on the right track.

File: src/Rasa.Game/Structures/ClanData.cs [hunk 1]
```diff
@@ -22,13 +26,29 @@ public void Read(PythonReader pr)
...
+            IsPvP = pr.ReadBool();
         }
 
         public void Write(PythonWriter pw)
         {
-            pw.WriteTuple(2);
+            pw.WriteTuple(4);
             pw.WriteUInt(Id);
             pw.WriteUnicodeString(Name);
+            pw.WriteTuple(RankTitles.Count);
```
Comment (krssrb) [inline src/Rasa.Game/Structures/ClanData.cs]: this shoiuld be pw.WriteList(RankTitles.Count);

File: src/Rasa.Game/Packets/MapChannel/Server/SetClanDataPacket.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            Name = name;
+            Data = data;
+        }
+
+        public override void Write(PythonWriter pw)
+        {
+            pw.WriteTuple(2);
+            pw.WriteString(Name);
+            pw.WriteTuple(4);
```
Comment (krssrb) [inline src/Rasa.Game/Packets/MapChannel/Server/SetClanDataPacket.cs]: ClanData is  IPythonDataStruct, so insted of `
pw.WriteTuple(4);
            pw.WriteString(Data.Name);
            pw.WriteUInt(Data.Id);

             foreach (string rankTitle in Data.RankTitles)
            {
                pw.WriteString(rankTitle);
            }

             pw.WriteBool(Data.IsPvP);`
you can use 
`pw.WriteStruct(ClanData);`

Comment (krssrb) [inline src/Rasa.Game/Packets/MapChannel/Server/SetClanDataPacket.cs]: ClanData is  IPythonDataStruct so insted of this
`pw.WriteTuple(4);
            pw.WriteString(Data.Name);
            pw.WriteUInt(Data.Id);`

            ` foreach (string rankTitle in Data.RankTitles)
            {
                pw.WriteString(rankTitle);
            }

             pw.WriteBool(Data.IsPvP);`
us can use `pw.WriteStruct(ClanData);`

File: src/Rasa.Game/Structures/ClanMemberData.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+﻿namespace Rasa.Structures
+{
+    public class ClanMemberData
```
Comment (krssrb) [inline src/Rasa.Game/Structures/ClanMemberData.cs]: you can change this class to `public class ClanMemberData : IPythonDataStruct` and add read and write functions and then use is with`pw.WriteStruct(ClanMemberData);` in packets `SetClanMemberDataPacket PlayerJoinedClanPacket` and others.
Look at ClanData class for example.

**Pull request merged**